src/ducc0/infra/threading.cc

```diff
@@ -607,7 +619,39 @@ void Distribution::thread_map(std::function<void(Scheduler &)> f)
   // threads, which executes everything sequentially on its own thread,
   // automatically prohibiting nested parallelism.
   auto pool = get_active_pool();
-  for (size_t i=0; i<nthreads_; ++i)
+
+#if 1  // Hierarchical submission
+
+  latch counter(nthreads_);
+  // distribute work to helper threads, in a recursive fashion
+  std::function<void(size_t, size_t)> new_f = [this, &f, &new_f, &counter, &ex, &ex_mut, pool](size_t istart, size_t step) {
```

Comment:
NIT: does this need to be a `std::function`?  It seems like it ought to just be a lambda, but then again I've never tried a self-referential lambda like this so maybe not.

Reply:
If I use `auto`, I get
```
./src/ducc0/infra/threading.cc:627:28: error: use of ‘new_f’ before deduction of ‘auto’
  627 |   auto new_f = [this, &f, &new_f, &counter, &ex, &ex_mut, pool](size_t istart, size_t step) {
      |                            ^~~~~
```

I think an explicit type is needed, and I'm not sure I can do better than using `std::function` here. The overhead should not be too large, I hope.

Reply:
You need to use a y-combinator for this.  We have something like that here: https://github.com/tensorflow/tensorflow/blob/6078149427d44b98f826cf9e7ef7f2761f6986a2/tensorflow/core/kernels/matmul_op_impl.h#L364